src/plugins/discover/public/application/main/discover_main_route.tsx

```diff
@@ -86,115 +84,74 @@ export function DiscoverMainRoute(props: Props) {
-        }
-
-        let defaultDataView: DataView | null = null;
-        try {
-          defaultDataView = await data.dataViews.getDefaultDataView({ displayErrors: false });
-        } catch (e) {
-          //
-        }
-
-        if (!defaultDataView) {
-          setShowNoDataPage(true);
-          return;
-        }
-
-        const { appState } = getDiscoverStateContainer({
-          history,
-          savedSearch: nextSavedSearch,
-          services,
-        });
-        const { index, query } = appState.getState();
-        const ip = await loadDataView(
-          data.dataViews,
-          config,
-          index,
-          historyLocationState?.dataViewSpec
-        );
-
-        const ipList = ip.list;
-        const isTextBasedQuery = query && isOfAggregateQueryType(query);
-        const dataViewData = resolveDataView(
-          ip,
-          nextSavedSearch.searchSource,
-          toastNotifications,
-          isTextBasedQuery
-        );
-        setDataViewList(ipList);
+  const checkData = useCallback(async () => {
+    try {
+      const hasUserDataViewValue = await data.dataViews.hasData
+        .hasUserDataView()
+        .catch(() => false);
+      const hasESDataValue = isDev || (await data.dataViews.hasData.hasESData().catch(() => false));
+      setHasUserDataView(hasUserDataViewValue);
+      setHasESData(hasESDataValue);
+
+      if (!hasUserDataViewValue) {
+        setShowNoDataPage(true);
+        return false;
+      }
 
-        return dataViewData;
+      let defaultDataView: DataView | null = null;
+      try {
+        defaultDataView = await data.dataViews.getDefaultDataView({ displayErrors: false });
       } catch (e) {
-        setError(e);
+        //
       }
-    },
-    [
-      config,
-      data.dataViews,
-      history,
-      isDev,
-      historyLocationState?.dataViewSpec,
-      toastNotifications,
-      services,
-    ]
-  );
+
+      if (!defaultDataView) {
+        setShowNoDataPage(true);
+        return false;
+      }
+      return true;
+    } catch (e) {
+      setError(e);
+      return false;
+    }
+  }, [data.dataViews, isDev]);
 
   const loadSavedSearch = useCallback(
     async (nextDataView?: DataView) => {
+      setLoading(true);
+      if (!nextDataView && !(await checkData())) {
+        setLoading(false);
+        return;
+      }
       try {
-        setLoading(true);
-        const currentSavedSearch = await getSavedSearch(id, {
-          search: services.data.search,
-          savedObjectsClient: core.savedObjects.client,
-          spaces: services.spaces,
-          savedObjectsTagging: services.savedObjectsTagging,
+        await stateContainer.actions.loadDataViewList();
+        // reset appState in case a saved search with id is loaded and the url is empty
+        // so the saved search is loaded in a clean state
+        // else it might be updated by the previous app state
+        const useAppState = !stateContainer.appState.isEmptyURL();
```

Comment:
How about 

```
// There are 2 ways the state in URL is handled
// 1. Loading saved searches or creating new ones in the UI
// New: `app/discover#/`
// View: `app/discover#/view/3ba638e0-b894-11e8-a6d9-e546fe2bba5`
// Here there's no state in URL.  In this case AppState has not to be considered when loading the saved search
// 
// 2. Any case when there's `_g` .. global State or `_a` appState involved with this URLs, then AppState has to be considered to update the given savedSearch
```

Reply:
> I don't think I understand what's going on here or the comment. Can I ask for the long form answer on this one? 😅
> 
> Also, since `useAppState` is checking the state container anyway, should we put it within `loadSavedSearch` instead?

This could make sense, I think this area could benefit from further improvements. the initial motivation here was to keep out URL state to simplify the unit tests. I'd suggest to fine tune in a follow up, given the actual size of this 👹 PR